udf-compiler/src/main/scala/com/nvidia/spark/udf/CatalystExpressionBuilder.scala

```diff
@@ -176,7 +176,36 @@ case class CatalystExpressionBuilder(private val function: AnyRef) extends Loggi
     // A basic block can have other branching instructions as the last instruction,
     // otherwise.
     if (basicBlock.lastInstruction.isReturn) {
-      newStates(basicBlock).expr
+      // When the compiled method throws exceptions, we combine the returned
```

Comment:
correct me if I am wrong, the test we have doesn't show this collapse behavior. I am hoping we can get a test that shows this.

Reply:
actually, `test("throw a SparkException object")` tests it.
The following analyzed logical plan is produced from `run(30)`:

```
== Analyzed Logical Plan ==
value: int, new: boolean
Project [value#4297, if (((value#4297 < 0) OR (value#4297 >= 30))) cast(raise_error(concat(concat(concat(concat(concat(, Fold number must be in range [0, ), cast(30 as string)), ), but
got ), cast(value#4297 as string)), .), NullType) as boolean) else true AS new#4300]
+- LocalRelation [value#4297]
```

Reply:
Ok, I also tried a more complex example:
```
      val myudf: (Int) => Boolean = i => {
        if (i < 0 || i >= x) {
          throw new SparkException(s"Fold number must be in range [0, $x), but got $i.")
        } else if (x % 2 != 0) {
          throw new SparkException(s"Invalid fold range! [0, $x). End of the range must be an even number.")
        }
        true
      }
``` 

That produced (with `run(30)`):
```
Project [value#5, if (((value#5 < 0) OR (value#5 >= 30))) cast(raise_error(concat(concat(concat(concat(concat(, Fold number must be in range [0, ), cast(30 as string)), ), but got ), cast(value#5 as string)), .), NullType) as boolean) else if ((((value#5 >= 0) AND (value#5 < 30)) AND NOT ((30 % 2) = 0))) cast(raise_error(concat(concat(concat(, Invalid fold range! [0, ), cast(30 as string)), ). End of the range must be an even number.), NullType) as boolean) else true AS my_fn(value)#9]
```

Which seems correct, though it does check the negated boolean expression of the if branch.